### PR TITLE
[AIRFLOW-161] New redirect route and extra links

### PR DESCRIPTION
**Pull request by ArgentFalcon**
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-161

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:

This PR is an attempt to finish the work started in #2657

With this change different operators would be able to customize the
task instance model view with extra links, those can be used to
redirect users to systems which are out of Airflow.

In order to be able to display the link on UI, and not have the backend do the external routing, I had to setup the endpoint to
be a RESTful endpoint that the frontend could ping whenever. I also wanted to handle a handful of error cases, so it returns different errors
If the URL is not whitelisted, if the URL is not provided, and there is even a way for operator to provide their own custom errors

Update the modal box to make a AJAX call for link resolutions
This gives us multiple benefits:
 1. It enables us to disable links when actionable
 2. It enables us to give better feedback to the user on whitelisting and errors without navigating the page
 3. You can see and interact with the link as it points to it's real destination, as opposed to the airflow backend doing the routing
   This let's you see where the link will resolve in advance

**Note**: A lot of the changes here are duplicating UI work between `www` and `www_rbac`

New additions: 
You can now see where the link will resolve in advance of clicking it:
screenshot 2018-06-21 18 33 22

Domains that aren't whitelisted will show a tooltip and be disabled
screenshot 2018-06-21 18 33 36

An operator can set a custom error message to show up on a tooltip (for example, if a link is not yet available for some reason, e.g. A job hasn't run yet, so no logs exist)
screenshot 2018-06-21 18 33 42

I should also mention here that at Lyft, we've had this running in production for half a year, and it's been really useful so far. 

I will also make a follow up PR with some improvements we've made to extra_links, allowing the user to specify extra_links as plugins that will apply to all operators. We use this feature to link to some logging tools we use (Kibana)

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
tests/www/test_views.py::TestRedirect
tests/www_rbac/test_views.py::TestRedirect
tests/contrib/operators/test_qubole_operator.py::QuboleOperatorTest::test_get_redirect_url
tests/models.py::DagTest::test_extra_links_no_affect
tests/models.py::DagTest:: test_extra_links

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by ArgentFalcon** [comment]
@mistercrunch PTAL

File: airflow/models.py [hunk 1]
```diff
@@ -3022,6 +3024,21 @@ def xcom_pull(
             dag_id=dag_id,
             include_prior_dates=include_prior_dates)
 
+    def get_redirect_url(self, dttm, redirect_to):
```
Comment (astahlman) [inline airflow/models.py]: How about just `get_extra_links()`?

"redirect" calls to mind the server returning an HTTP 3xx, which this is not.

File: airflow/www/views.py [hunk 2]
```diff
@@ -1893,6 +1910,83 @@ def gantt(self, session=None):
             root=root,
         )
 
+    @expose('/redirect')
+    @login_required
+    @wwwutils.action_logging
+    def redirect(self):
```
Comment (astahlman) [inline airflow/www/views.py]: Similar comment as above re: the use of the term "redirect" - this really just returns a list of URLs (or "extra_links", as they are called elsewhere.)

File: airflow/www/views.py [hunk 3]
```diff
@@ -1893,6 +1910,83 @@ def gantt(self, session=None):
...
+            return response
+        if url:
+            allowed_domains = conf.get('webserver', 'whitelisted_domains')
+            parsed_uri = urlparse(url)
+            domain = '{uri.scheme}://{uri.netloc}'.format(uri=parsed_uri)
+            if domain in allowed_domains:
+                response = jsonify({'error': None,
+                                    'url': url})
+                response.status_code = 200
+                print(response)
```
Comment (astahlman) [inline airflow/www/views.py]: Is this print statement leftover from debugging?

**Review by astahlman — COMMENTED** [review]
A couple of points to consider:

1. Do we still need the domain whitelist? I think that may be an artifact from when an earlier version of this PR issued HTTP redirects. Now that the server doesn't issue redirects, I'm not sure this whitelist is still valuable.

2. Rather than fetching this list of "extra links" via a client-side AJAX call, why not just serve the page with these links pre-populated? I think it would greatly simplify this implementation if this rendering happened server-side. The only advantage I see to populating them client-side is that the links can be dynamically updated without refreshing the page. The downside is that it adds a lot of complexity to the client-side rendering code (e.g., dealing with non 2xx responses from the `/redirect` endpoint). Personally, I don't think the increased responsiveness justifies the additional complexity.

File: airflow/models.py (hunk 1 above)
Comment (ArgentFalcon) [reply airflow/models.py]: Good point, can change this.

**Comment by ArgentFalcon** [comment]
Okay, after observing my behavior with the operator, I refresh a lot anyways to observe the state, so I think it does make think to move it from AJAX call to having the links pre-populated.

File: airflow/contrib/hooks/qubole_hook.py [hunk 4]
```diff
@@ -182,6 +185,29 @@ def get_jobs_id(self, ti):
...
         Command.get_jobs_id(self.cls, cmd_id)
 
+    def get_redirect_url(self, task, dttm):
+        session = settings.Session()
+        url = ''
+
+        try:
+            conn = BaseHook.get_connection(task.kwargs['qubole_conn_id'])
+            if conn and conn.host:
+                host = re.sub(r'api$', 'v2/analyze?command_id=', conn.host)
```
Comment (feng-tao) [inline airflow/contrib/hooks/qubole_hook.py:214]: is v2/analyze?command_id= generic used for qubole hook?

File: airflow/contrib/hooks/qubole_hook.py [hunk 5]
```diff
@@ -182,6 +185,29 @@ def get_jobs_id(self, ti):
...
+            else:
+                host = 'https://api.qubole.com/v2/analyze?command_id='
+
+            ti = TaskInstance(task=task, execution_date=dttm)
+            qds_command_id = ti.xcom_pull(task_ids=task.task_id, key='qbol_cmd_id')
+
+            url = host + str(qds_command_id) if qds_command_id else ''
+        except Exception as e:
+            print('Could not find the url to redirect. Error: %s' % str(e))
```
Comment (feng-tao) [inline airflow/contrib/hooks/qubole_hook.py]: should use LoggingMixing method instead or print

File: airflow/contrib/operators/qubole_operator.py [hunk 6]
```diff
@@ -160,6 +161,9 @@ def get_hook(self):
         # Reinitiating the hook, as some template fields might have changed
         return QuboleHook(*self.args, **self.kwargs)
 
+    def get_redirect_url(self, dttm, redirect_to):
+        return self.get_hook().get_redirect_url(self, dttm)
```
Comment (feng-tao) [inline airflow/contrib/operators/qubole_operator.py]: the `get_redirect_url` in qubole hook has task argument?

**Review by feng-tao — COMMENTED** [review]
will look at UI code later

**Comment by ArgentFalcon** [comment]
After trying to do it statically, I ran into the same issue @msumit did with the tree view. The tree view of a DAG loads up a bunch of dates at once, that complicates the logic decently. I think the AJAX calls is actually a simpler solution for the moment.

File: airflow/contrib/operators/qubole_operator.py (hunk 6 above)
Comment (ArgentFalcon) [reply airflow/contrib/operators/qubole_operator.py]: Yes. It pulls from xcom, which is task based, so it needs the task instance to do that.

**Comment by ArgentFalcon** [comment]
@astahlman removed the whiltelist

File: airflow/www/views.py [hunk 7]
```diff
@@ -1893,6 +1910,70 @@ def gantt(self, session=None):
...
+        redirect_to = request.args.get('redirect_to')
+        dttm = airflow.utils.timezone.parse(execution_date)
+        dag = dagbag.get_dag(dag_id)
+
+        if not dag or task_id not in dag.task_ids:
+            flash(
+                "Task [{}.{}] doesn't seem to exist"
+                " at the moment".format(dag_id, task_id),
+                "error")
+            return redirect(request.referrer or '/admin/')
```
Comment (msumit) [inline airflow/www/views.py]: `flash` and `redirect` should not be part of an ajax api

File: airflow/contrib/hooks/qubole_hook.py (hunk 4 above)
Comment (msumit) [reply airflow/contrib/hooks/qubole_hook.py:214]: yes

File: airflow/www/views.py (hunk 7 above)
Comment (ArgentFalcon) [reply airflow/www/views.py]: fixed. Thanks for the catch :)

File: airflow/www_rbac/views.py [hunk 8]
```diff
@@ -1705,6 +1715,68 @@ def gantt(self, session=None):
             root=root,
         )
 
+    @expose('/extra_links')
+    @has_access
```
Comment (feng-tao) [inline airflow/www_rbac/views.py]: need to add @has_dag_access something decorator for DAG level access

**Comment by ron819** [comment]
@ArgentFalcon are you still working on this PR?

**Comment by ArgentFalcon** [comment]
Oh yeah, I should finish this up. I have to pull some more changes that we did internally at Lyft that make it more versatile.

**Comment by ArgentFalcon** [comment]
Maybe I should just fix the commits and merge it and then make a new PR with more changes. I'll do that.

**Comment by msumit** [comment]
@ArgentFalcon maybe you want to update the screenshots and description of the PR as well?

File: airflow/www_rbac/views.py (hunk 8 above)
Comment (ArgentFalcon) [reply airflow/www_rbac/views.py]: Cool, will add.

**Comment by ArgentFalcon** [comment]
@msumit The screenshots are still accurate. I updated the descriptions with our experience with this improvement, as well a follow up PR. Most of the code here is UI code, which would be nice to get merged, as that doesn't change

@feng-tao, fixed conflicts

File: airflow/contrib/hooks/qubole_hook.py [hunk 9]
```diff
@@ -200,6 +203,30 @@ def get_jobs_id(self, ti):
             cmd_id = ti.xcom_pull(key="qbol_cmd_id", task_ids=self.task_id)
         Command.get_jobs_id(self.cls, cmd_id)
 
+    def get_extra_links(self, task, dttm):
```
Comment (feng-tao) [inline airflow/contrib/hooks/qubole_hook.py:206]: docstring

File: airflow/contrib/operators/qubole_operator.py [hunk 10]
```diff
@@ -174,6 +175,9 @@ def get_hook(self):
         # Reinitiating the hook, as some template fields might have changed
         return QuboleHook(*self.args, **self.kwargs)
 
+    def get_extra_links(self, dttm, link_name):
```
Comment (feng-tao) [inline airflow/contrib/operators/qubole_operator.py:182]: docstring

File: airflow/www_rbac/views.py [hunk 11]
```diff
@@ -1772,6 +1782,68 @@ def gantt(self, session=None):
             root=root,
         )
 
+    @expose('/extra_links')
+    @has_dag_access
```
Comment (feng-tao) [inline airflow/www_rbac/views.py]:
```
@has_dag_access(can_dag_read=True)
@has_access
@action_logging
```

**Review by feng-tao — APPROVED** [review]
some small nits.

File: airflow/www_rbac/views.py (hunk 11 above)
Comment (ArgentFalcon) [reply airflow/www_rbac/views.py]: Thanks.

File: airflow/contrib/hooks/qubole_hook.py (hunk 9 above)
Comment (ArgentFalcon) [reply airflow/contrib/hooks/qubole_hook.py:206]: This is an implementation of https://github.com/apache/airflow/pull/3533/files#diff-664e0133fa172b6e0060f2dd7b2c3504R2948, which does have a docstring. Is that not sufficient?

File: airflow/contrib/operators/qubole_operator.py (hunk 10 above)
Comment (ArgentFalcon) [reply airflow/contrib/operators/qubole_operator.py:182]: This is an implementation of https://github.com/apache/airflow/pull/3533/files#diff-664e0133fa172b6e0060f2dd7b2c3504R2948

**Comment by feng-tao** [comment]
@ArgentFalcon , let me know if you finish the final update. I will do a final pass.

**Comment by feng-tao** [comment]
And please squash the commit into a single one.

**Comment by feng-tao** [comment]
@ArgentFalcon CI fails.

**Comment by ron819** [comment]
The failure doesn't seem to be related to the PR?

ERROR [airflow.models.DagBag] Failed to import: /app/tests/dags/test_impersonation_custom.py
Traceback (most recent call last):
  File "/app/airflow/models/__init__.py", line 382, in process_file
    m = imp.load_source(mod_name, filepath)
  File "/app/tests/dags/test_impersonation_custom.py", line 31, in <module>
    from tests.test_utils.fake_datetime import FakeDatetime
  File "/app/tests/__init__.py", line 26, in <module>
    from .models import *
  File "/app/tests/models.py", line 53, in <module>
    from airflow.operators import BaseOperator
ImportError: cannot import name BaseOperator

**Comment by feng-tao** [comment]
@ArgentFalcon  Airflow just deleted all the old UI code. So you need to rebase your change and only modify it with new UI.

**Comment by ashb** [comment]
Does this show up on the TaskInstance pages, or just the pop up?

(I think anything that is in the pop-up should also be visible from the task instance pages as well)

**Comment by feng-tao** [comment]
@ashb , the pr provides a way to show extra_link as plugin and display in task instance pop up if there is one. Hence I am not sure if it is a good idea to show in task instance page as well as people could have no extra link.

**Comment by feng-tao** [comment]
But @ArgentFalcon , I  don't think this pr could make the 10.2 release as Kaxil will cut the release in these few days. cc @astahlman

**Comment by kaxil** [comment]
This is some nice work @ArgentFalcon and indeed would be a very useful feature for the community.  It would great if you can find some time and rebase it to master.

File: airflow/contrib/hooks/qubole_hook.py [hunk 12]
```diff
@@ -200,6 +203,30 @@ def get_jobs_id(self, ti):
...
+                host = re.sub(r'api$', 'v2/analyze?command_id=', conn.host)
+            else:
+                host = 'https://api.qubole.com/v2/analyze?command_id='
+
+            ti = TaskInstance(task=task, execution_date=dttm)
+            qds_command_id = ti.xcom_pull(task_ids=task.task_id, key='qbol_cmd_id')
+
+            url = host + str(qds_command_id) if qds_command_id else ''
+        except Exception as e:
+            log.warning('Could not find the url to redirect. Error: %s' % str(e))
```
Comment (mik-laj) [inline airflow/contrib/hooks/qubole_hook.py:223]:
```suggestion
            log.warning('Could not find the url to redirect. Error: %s', e)
```
You should avoid format message before pass to logger.

**Comment by msumit** [comment]
@ArgentFalcon I think we should be good to merge this PR, once Travis turns green.

**Comment by feng-tao** [comment]
CI still fails

**Comment by ArgentFalcon** [comment]
@feng-tao I did it. The tests pass

**Comment by ArgentFalcon** [comment]
nevermind
as soon as I typed that a failure showed up :(

**Comment by feng-tao** [comment]
thanks @ArgentFalcon , lgtm

**Comment by feng-tao** [comment]
@ArgentFalcon , somehow the test is consistently failing in master branch. I am going to revert it for now. Please investigate and resubmit the pr :(

**Comment by feng-tao** [comment]
@ArgentFalcon this pr has been reverted. Do let me know if you have the chance to resubmit with fixing all the test failures.

**Comment by feng-tao** [comment]
@ArgentFalcon , This pr has been reverted due to unit test failure. not sure if you have time to revisit the pr.
cc @milton0825 @astahlman and see if you two have cycles to finish the pr.

**Pull request merged**